src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import akka.http.scaladsl.model.StatusCodes
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.plus.PlusScopes
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.gson.internal.LinkedTreeMap
+import com.typesafe.config.Config
+import io.swagger.client.ApiClient
+import io.swagger.client.api.ResourcesApi
+import io.swagger.client.api.GoogleApi
+import java.io.{ByteArrayInputStream, File}
+import java.util.concurrent.TimeUnit
+import com.google.gson.Gson
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model._
+import org.broadinstitute.dsde.workbench.leonardo.model.Actions._
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions._
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions._
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import net.ceedubs.ficus.Ficus._
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+case class UserEmailAndProject(userEmail: String, googleProject: String)
+
+case class NotebookActionError(action: Action) extends
+  LeoException(s"${action.toString} was not recognized", StatusCodes.NotFound)
+
+
+class SamAuthProvider(authConfig: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(authConfig, serviceAccountProvider) with LazyLogging {
+
+  private val notebookClusterResourceTypeName = "notebook-cluster"
+  private val billingProjectResourceTypeName = "billing-project"
+
+  private val httpTransport = GoogleNetHttpTransport.newTrustedTransport
+  private val jsonFactory = JacksonFactory.getDefaultInstance
+  private val saScopes = Seq(PlusScopes.USERINFO_EMAIL, PlusScopes.USERINFO_PROFILE)
+
+  //Cache lookup of pet tokens
+  val cacheExpiryTime = authConfig.getInt("cacheExpiryTime")
+  val cacheMaxSize = authConfig.getInt("cacheMaxSize")
+
+  private[leonardo] val petTokenCache = CacheBuilder.newBuilder()
```

Comment:
This cache will go away once we start using the pet to call Sam, right?

Reply:
this _is_ using the pet to call Sam! it uses leo'd admin sam-auth to grab the pet's creds and construct tokens from that.

Reply:
we could ditch this cache but we'd be repeatedly going to sam to yank the json and create the credential every time we need a pet cred.

Reply:
Oh geez, comprehension fail. So this:

1. calls a Sam admin endpoint *as leo* passing in a (userEmail, gProj) to get a pet token
2. caches the results keyed by (userEmail, gproj)
3. when a needed, looks up the pet token from cache and calls Sam with the pet token
4. Sam switches back to the user's identity

Reply:
yup! it's a big, funny loop, but there's a certain elegance to how it all works together. switcheroo is magic

Reply:
Question: is Sam managing how often it's creating service account keys?  I.e. whenever the cache expires (or Leo starts up) we're going to call Sam `/admin/getUserPetServiceAccountKey`. I hope that isn't going to cause Sam to create a _new_ key in Google.

Reply:
yeah, sam is handling that. i think they're doing the stupid thing at first and then will roll out less stupid approaches in future. as far as we're concerned this is [SEP](https://en.wikipedia.org/wiki/Somebody_else%27s_problem)

Reply:
I should probably add comments in the code to explain stuff